lazy_load_symbols/load_sym_lazily.py

```diff
@@ -0,0 +1,255 @@
+    symc = re.compile(symre)
+    for line in syml.split('\n'):
+        m = symc.search(line)
+        if m:
+            yield(m.group(1))
+
+
+def extract_symbol_list_from_so(file_name):
+    try:
+        symlist = subprocess.check_output(['nm', '-C', file_name])
+    except subprocess.CalledProcessError:
+        print('nm -C {} failed'.format(file_name))
+        raise
+    except OSError:
+        print('no usable "nm" found, cannot continue')
+        raise
+    assert symlist, 'no symbols in file {}'.format(file_name)
+    return symlist
+
+
+def extract_symbols_from_raw_list(rs):
+    '''
+    Saves the entire symbol names as a single long string
+    so we can say something like:
+        if symbol in symbol_list
+    '''
+    parsed_list = []
+    for sym in rs:
+        sym_name = subprocess.check_output(['c++filt', sym])
+        assert sym_name, 'c++filt didn\'t demangle {}'.format(sym)
+        parsed_list.append(sym_name)
+    return ''.join(parsed_list)
+
+SOInfo = namedtuple('SOInfo', ['fullname', 'symbols', 'start', 'end', 'loaded'])
+
+def load_so_list():
+    '''
+    Loads the list of shared libraries and returns a dictionary
+    '''
+    solist = gdb.execute('info sharedlibrary', to_string=True)
+    solibre = r'0x([0-9a-f]+)\s+0x([0-9a-f]+).+ ([^\0]+)\/(.+)'
+    solibc = re.compile(solibre)
+    sodict = {}
+    for so in solist.splitlines():
+        m = solibc.match(so)
+        if m:
+            name = m.group(4)
+            so_path = m.group(3)
+            sa = m.group(1)
+            ea = m.group(2)
+            if name not in sodict:
+                sodict[name] = SOInfo(
+                    fullname = so_path + '/' + name,
+                    symbols = {},
+                    start = sa,
+                    end = ea,
+                    loaded = False
+                    )
+            else:
+                # this shouldn't happen but we can just verify it is indeed the same library
+                assert sodict[name].start == sa, 'found same' \
+                    ' library ({}) at 2 different locations: {} != {}' \
+                    .format(name, sa, sodict[name].start)
+                assert sodict[name].end == ea, 'found same library' \
+                    ' ({}) ending at 2 different locations: {} != {}' \
+                    .format(name, ea, sodict[name].end)
+    return sodict
+
+class LoadLazySymbols(gdb.Command):
+    '''
+    A command to load a recording without loading
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMMAND_DATA, gdb.COMPLETE_FILENAME)
+        self.sodict = {}
+        self.rec = None
+
+
+    def load_all_symbols(self):
+        for content in self.sodict.values():
+            fn = content.fullname
+            content.symbols = extract_symbol_list_from_so(fn)
+
+    def invoke(self, rec, from_tty):
+        try:
+            if self.sodict and \
```

Comment:
If the user tries to load again, why not just loading again?

Reply:
it is time consuming, if it is the same recording we don't want to waste time. (consider the user accidentally hitting enter again)